src/components/LineChart/types.ts

```diff
@@ -0,0 +1,126 @@
+import { SizeMeProps } from 'react-sizeme';
+import { ItemId, Domain } from 'external';
+import { AxisPlacement } from 'components/AxisPlacement';
+import { AxisDisplayMode } from 'utils/AxisDisplayMode';
+
+type Point = {
+  id: ItemId;
+  name: string;
+  color: string;
+  timestamp: Date;
+  rawTimestamp?: Date;
+  value: number;
+  x: number;
+  y: number;
+};
+
+type Ruler = {
+  visible: boolean;
+  timeLabel: React.ReactNode;
+  yLabel: React.ReactNode;
+  // a timestamp representing the initial position of the ruler
+  timestamp: number;
+  // a function that determines the position of the timestamp
+  // (defaultPosition:number, {height:number, toolTipHeight:number, timeLabelMargin:number}) => number
+  getTimeLabelPosition: (
+    defaultPosition: number,
+    layout: { height: number; toolTipHeight: number; timeLabelMargin: number }
+  ) => number;
+};
+
+type ContextChart = {
+  visible: boolean;
+  height: number;
+  isDefault: boolean;
+};
+
+type Annotation = {
+  data: number[];
+  xScale: (rawX: number) => number;
+  height: number;
+  id: ItemId;
+  color: string;
+  fillOpacity: number;
+};
+
+type TickFormatter = (value: number, values: number[]) => number | string;
+
+type Coordinate = {
+  xval: number;
+  yval: number;
+  points: Point[];
+};
+
+type Area = {
+  id: ItemId;
+  color: string;
+  start?: Coordinate;
+  end?: Coordinate;
+};
+
+export type LineChartProps = SizeMeProps & {
+  // size props
+  width?: number;
+  height?: number;
+
+  // visual options
+  zoomable?: boolean;
+  crosshair?: boolean;
+  pointWidth?: number;
+
+  xSubDomain?: Domain;
+  xAxisHeight?: number;
+  xAxisFormatter?: TickFormatter;
+  xAxisPlacement?: AxisPlacement;
+
+  yAxisWidth?: number;
+  yAxisTicks?: number;
+  yAxisFormatter?: TickFormatter;
+  yAxisDisplayMode?: AxisDisplayMode;
+
+  // additional display elements
+  contextChart?: ContextChart;
+  ruler?: Ruler;
+  annotations?: Annotation[];
+  areas?: Area[];
+
+  // event based function props
+  onMouseMove?: (args: {
+    points: Point[];
+    xpos?: number;
+    ypos?: number;
+    e: React.MouseEvent;
+  }) => void;
+  onMouseOut?: (e: React.MouseEvent) => void;
+  onBlur?: (e: React.MouseEvent) => void;
+  onClick?: (e: React.MouseEvent) => void;
+  // (annotation, x, y) => void
+  // onClickAnnotation: func,
+  // event => void
+  // onDoubleClick: func,
+  // ({ xSubDomain, transformation }) => void
+  // onZoomXAxis: func,
+  // Number => String
+  // (e, seriesId) => void
+  // onAxisMouseEnter: func,
+  // (e, seriesId) => void
+  // onAxisMouseLeave: func,
```

Comment:
I haven't typed all the functions yet. The tests I'm creating right now won't use them (though they should be tested separately).